// @celo/contractkit/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !exports.hasOwnProperty(p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newWeb3 = void 0;
var web3_1 = __importDefault(require("web3"));
var base_1 = require("./base");
Object.defineProperty(exports, "AllContracts", { enumerable: true, get: function () { return base_1.AllContracts; } });
Object.defineProperty(exports, "CeloContract", { enumerable: true, get: function () { return base_1.CeloContract; } });
Object.defineProperty(exports, "NULL_ADDRESS", { enumerable: true, get: function () { return base_1.NULL_ADDRESS; } });
var identity_1 = require("./identity");
Object.defineProperty(exports, "IdentityMetadataWrapper", { enumerable: true, get: function () { return identity_1.IdentityMetadataWrapper; } });
__exportStar(require("./kit"), exports);
var genesis_block_utils_1 = require("./network-utils/genesis-block-utils");
Object.defineProperty(exports, "GenesisBlockUtils", { enumerable: true, get: function () { return genesis_block_utils_1.GenesisBlockUtils; } });
var static_node_utils_1 = require("./network-utils/static-node-utils");
Object.defineProperty(exports, "StaticNodeUtils", { enumerable: true, get: function () { return static_node_utils_1.StaticNodeUtils; } });
var celo_provider_1 = require("./providers/celo-provider");
Object.defineProperty(exports, "CeloProvider", { enumerable: true, get: function () { return celo_provider_1.CeloProvider; } });
var phone_number_lookup_utils_1 = require("./utils/phone-number-lookup/phone-number-lookup-utils");
Object.defineProperty(exports, "PNPUtils", { enumerable: true, get: function () { return phone_number_lookup_utils_1.PNPUtils; } });
var BaseWrapper_1 = require("./wrappers/BaseWrapper");
Object.defineProperty(exports, "CeloTransactionObject", { enumerable: true, get: function () { return BaseWrapper_1.CeloTransactionObject; } });
/**
 * Creates a new web3 instance
 * @param url node url
 */
function newWeb3(url) {
    return new web3_1.default(url);
}
exports.newWeb3 = newWeb3;
