// @bitgo/statics/dist/src/errors.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConflictingCoinFeaturesError = exports.InvalidDomainError = exports.InvalidContractAddressError = exports.MissingRequiredCoinFeatureError = exports.DisallowedCoinFeatureError = exports.DuplicateCoinDefinitionError = exports.CoinNotDefinedError = void 0;
var BitGoStaticsError = /** @class */ (function (_super) {
    __extends(BitGoStaticsError, _super);
    function BitGoStaticsError(message) {
        var _this = _super.call(this, message) || this;
        _this.stack = new Error(message).stack;
        Object.setPrototypeOf(_this, BitGoStaticsError.prototype);
        return _this;
    }
    return BitGoStaticsError;
}(Error));
var CoinNotDefinedError = /** @class */ (function (_super) {
    __extends(CoinNotDefinedError, _super);
    function CoinNotDefinedError(coinName) {
        var _this = _super.call(this, "coin '" + coinName + "' is not defined") || this;
        Object.setPrototypeOf(_this, CoinNotDefinedError.prototype);
        return _this;
    }
    return CoinNotDefinedError;
}(BitGoStaticsError));
exports.CoinNotDefinedError = CoinNotDefinedError;
var DuplicateCoinDefinitionError = /** @class */ (function (_super) {
    __extends(DuplicateCoinDefinitionError, _super);
    function DuplicateCoinDefinitionError(coinName) {
        var _this = _super.call(this, "coin '" + coinName + "' is already defined") || this;
        Object.setPrototypeOf(_this, DuplicateCoinDefinitionError.prototype);
        return _this;
    }
    return DuplicateCoinDefinitionError;
}(BitGoStaticsError));
exports.DuplicateCoinDefinitionError = DuplicateCoinDefinitionError;
var DisallowedCoinFeatureError = /** @class */ (function (_super) {
    __extends(DisallowedCoinFeatureError, _super);
    function DisallowedCoinFeatureError(coinName, feature) {
        var _this = _super.call(this, "coin feature '" + feature + "' is disallowed for coin " + coinName + ".") || this;
        Object.setPrototypeOf(_this, DisallowedCoinFeatureError.prototype);
        return _this;
    }
    return DisallowedCoinFeatureError;
}(BitGoStaticsError));
exports.DisallowedCoinFeatureError = DisallowedCoinFeatureError;
var MissingRequiredCoinFeatureError = /** @class */ (function (_super) {
    __extends(MissingRequiredCoinFeatureError, _super);
    function MissingRequiredCoinFeatureError(coinName, missingFeatures) {
        var _this = _super.call(this, "Required coin feature(s) '" + missingFeatures.join(', ') + "' were not found for coin " + coinName + ".") || this;
        Object.setPrototypeOf(_this, MissingRequiredCoinFeatureError.prototype);
        return _this;
    }
    return MissingRequiredCoinFeatureError;
}(BitGoStaticsError));
exports.MissingRequiredCoinFeatureError = MissingRequiredCoinFeatureError;
var InvalidContractAddressError = /** @class */ (function (_super) {
    __extends(InvalidContractAddressError, _super);
    function InvalidContractAddressError(coinName, contractAddress) {
        var _this = _super.call(this, "invalid contract address '" + contractAddress + "' for coin '" + coinName + "'") || this;
        Object.setPrototypeOf(_this, InvalidContractAddressError.prototype);
        return _this;
    }
    return InvalidContractAddressError;
}(BitGoStaticsError));
exports.InvalidContractAddressError = InvalidContractAddressError;
var InvalidDomainError = /** @class */ (function (_super) {
    __extends(InvalidDomainError, _super);
    function InvalidDomainError(coinName, domain) {
        var _this = _super.call(this, "invalid domain '" + domain + "' for coin '" + coinName + "'") || this;
        Object.setPrototypeOf(_this, InvalidDomainError.prototype);
        return _this;
    }
    return InvalidDomainError;
}(BitGoStaticsError));
exports.InvalidDomainError = InvalidDomainError;
var ConflictingCoinFeaturesError = /** @class */ (function (_super) {
    __extends(ConflictingCoinFeaturesError, _super);
    function ConflictingCoinFeaturesError(coinName, conflictingFeatures) {
        var _this = _super.call(this, "coin feature(s) for coin '" + coinName + "' cannot be both required and disallowed: " + conflictingFeatures.join(', ')) || this;
        Object.setPrototypeOf(_this, ConflictingCoinFeaturesError.prototype);
        return _this;
    }
    return ConflictingCoinFeaturesError;
}(BitGoStaticsError));
exports.ConflictingCoinFeaturesError = ConflictingCoinFeaturesError;
