// @bitgo/statics/dist/src/utxo.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.utxo = exports.UtxoCoin = void 0;
var base_1 = require("./base");
var UtxoCoin = /** @class */ (function (_super) {
    __extends(UtxoCoin, _super);
    function UtxoCoin(options) {
        var _this = _super.call(this, __assign(__assign({}, options), { kind: base_1.CoinKind.CRYPTO, isToken: false, decimalPlaces: 8 })) || this;
        _this.network = options.network;
        return _this;
    }
    UtxoCoin.prototype.disallowedFeatures = function () {
        return new Set([base_1.CoinFeature.ACCOUNT_MODEL]);
    };
    UtxoCoin.prototype.requiredFeatures = function () {
        return new Set([base_1.CoinFeature.UNSPENT_MODEL]);
    };
    UtxoCoin.DEFAULT_FEATURES = [
        base_1.CoinFeature.UNSPENT_MODEL,
        base_1.CoinFeature.CHILD_PAYS_FOR_PARENT,
        base_1.CoinFeature.WRAPPED_SEGWIT,
        base_1.CoinFeature.NATIVE_SEGWIT,
    ];
    return UtxoCoin;
}(base_1.BaseCoin));
exports.UtxoCoin = UtxoCoin;
/**
 * Factory function for utxo coin instances.
 *
 * @param name unique identifier of the coin
 * @param fullName Complete human-readable name of the coin
 * @param network Network object for this coin
 * @param asset Asset which this coin represents. This is the same for both mainnet and testnet variants of a coin.
 * @param features? Features of this coin. Defaults to the DEFAULT_FEATURES defined in `UtxoCoin`
 * @param prefix? Optional coin prefix. Defaults to empty string
 * @param suffix? Optional coin suffix. Defaults to coin name.
 * @param primaryKeyCurve The elliptic curve for this chain/token
 */
function utxo(name, fullName, network, asset, features, prefix, suffix, 
/** All UTXOs BitGo supports are SECP256K1 **/
primaryKeyCurve) {
    if (features === void 0) { features = UtxoCoin.DEFAULT_FEATURES; }
    if (prefix === void 0) { prefix = ''; }
    if (suffix === void 0) { suffix = name.toUpperCase(); }
    if (primaryKeyCurve === void 0) { primaryKeyCurve = base_1.KeyCurve.Secp256k1; }
    return Object.freeze(new UtxoCoin({
        name: name,
        fullName: fullName,
        network: network,
        prefix: prefix,
        suffix: suffix,
        features: features,
        asset: asset,
        primaryKeyCurve: primaryKeyCurve,
    }));
}
exports.utxo = utxo;
