// bitgo/dist/src/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @prettier
 */
//
// index.js - Module definition for BitGoJS
//
// Copyright 2019, BitGo, Inc.  All Rights Reserved.
//
var common = require("./common");
__export(require("./bitgo"));
// Expose bitcoin and sjcl
var utxoLib = require("@bitgo/utxo-lib");
var bitcoin_1 = require("./bitcoin");
// can't add types for these since they are part of @bitgo/utxo-lib's default export
// see https://github.com/Microsoft/TypeScript/issues/14080
utxoLib.hdPath = bitcoin_1.hdPath;
utxoLib.makeRandomKey = bitcoin_1.makeRandomKey;
exports.bitcoin = utxoLib;
exports.sjcl = require('./vendor/sjcl.min.js');
exports.bs58 = require('bs58');
var buffer_1 = require("buffer");
exports.Buffer = buffer_1.Buffer;
var _ = require("lodash");
var errors = require("./errors");
exports.Environments = _.cloneDeep(common.Environments);
exports.Errors = errors;
var coinFactory_1 = require("./v2/coinFactory");
exports.GlobalCoinFactory = coinFactory_1.GlobalCoinFactory;
__export(require("./v2"));
/**
 * Set the network, i.e. either "bitcoin" for production with real bitcoin, or
 * "testnet" for development with testnet bitcoin.
 *
 * @deprecated
 */
function setNetwork(network) {
    common.setNetwork(network);
}
exports.setNetwork = setNetwork;
/**
 * Get the network. Returns either "bitcoin" or "testnet".
 *
 * @deprecated
 */
function getNetwork() {
    return common.getNetwork();
}
exports.getNetwork = getNetwork;
/**
 * @deprecated
 */
function getNetworkObj() {
    return exports.bitcoin.networks[common.getNetwork()];
}
exports.getNetworkObj = getNetworkObj;
setNetwork('testnet');
