// bitgo/dist/src/v2/coins/bsv.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var bch_1 = require("./bch");
var bitcoin = require("@bitgo/utxo-lib");
var request = require('superagent');
var Bluebird = require("bluebird");
var co = Bluebird.coroutine;
var common = require("../../common");
var errors = require("../../errors");
var blockchairApi_1 = require("../recovery/blockchairApi");
var Bsv = /** @class */ (function (_super) {
    __extends(Bsv, _super);
    function Bsv(bitgo, network) {
        return _super.call(this, bitgo, network || bitcoin.networks.bitcoinsv) || this;
    }
    Bsv.createInstance = function (bitgo) {
        return new Bsv(bitgo);
    };
    Bsv.prototype.getChain = function () {
        return 'bsv';
    };
    Bsv.prototype.getFamily = function () {
        return 'bsv';
    };
    Bsv.prototype.getFullName = function () {
        return 'Bitcoin SV';
    };
    Bsv.prototype.recoveryBlockchainExplorerUrl = function (url) {
        var baseUrl = common.Environments[this.bitgo.getEnv()].bsvExplorerBaseUrl;
        // TODO BG-9989: There is no explorer api for Bitcoin SV yet. Once we have one, add it to src/common.js and update
        // this method.
        if (!baseUrl) {
            throw new errors.WalletRecoveryUnsupported("Recoveries not supported for " + this.getChain() + " - no explorer available");
        }
        return common.Environments[this.bitgo.getEnv()].bsvExplorerBaseUrl + url;
    };
    Bsv.prototype.getAddressInfoFromExplorer = function (addressBase58, apiKey) {
        var explorer = new blockchairApi_1.BlockchairApi(this.bitgo, 'bitcoin-sv', apiKey);
        return Bluebird.resolve(explorer.getAccountInfo(addressBase58));
    };
    Bsv.prototype.getUnspentInfoFromExplorer = function (addressBase58, apiKey) {
        var explorer = new blockchairApi_1.BlockchairApi(this.bitgo, 'bitcoin-sv', apiKey);
        return Bluebird.resolve(explorer.getUnspents(addressBase58));
    };
    return Bsv;
}(bch_1.Bch));
exports.Bsv = Bsv;
