// bitgo/dist/src/v2/coins/celoToken.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var celo_1 = require("./celo");
var statics_1 = require("@bitgo/statics");
var CeloToken = /** @class */ (function (_super) {
    __extends(CeloToken, _super);
    function CeloToken(bitgo, tokenConfig) {
        var _this = this;
        var staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('celo') : statics_1.coins.get('tcelo');
        _this = _super.call(this, bitgo, staticsCoin) || this;
        _this.tokenConfig = tokenConfig;
        return _this;
    }
    CeloToken.createTokenConstructor = function (config) {
        return function (bitgo) { return new CeloToken(bitgo, config); };
    };
    Object.defineProperty(CeloToken.prototype, "type", {
        get: function () {
            return this.tokenConfig.type;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CeloToken.prototype, "name", {
        get: function () {
            return this.tokenConfig.name;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CeloToken.prototype, "coin", {
        get: function () {
            return this.tokenConfig.coin;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CeloToken.prototype, "network", {
        get: function () {
            return this.tokenConfig.network;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CeloToken.prototype, "tokenContractAddress", {
        get: function () {
            return this.tokenConfig.tokenContractAddress;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CeloToken.prototype, "decimalPlaces", {
        get: function () {
            return this.tokenConfig.decimalPlaces;
        },
        enumerable: true,
        configurable: true
    });
    CeloToken.prototype.getChain = function () {
        return this.tokenConfig.type;
    };
    CeloToken.prototype.getFullName = function () {
        return 'Celo Token';
    };
    CeloToken.prototype.getBaseFactor = function () {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    };
    /**
     * Flag for sending value of 0
     * @returns {boolean} True if okay to send 0 value, false otherwise
     */
    CeloToken.prototype.valuelessTransferAllowed = function () {
        return false;
    };
    /**
     * Flag for sending data along with transactions
     * @returns {boolean} True if okay to send tx data (CELO), false otherwise
     */
    CeloToken.prototype.transactionDataAllowed = function () {
        return false;
    };
    return CeloToken;
}(celo_1.Celo));
exports.CeloToken = CeloToken;
