// bitgo/dist/src/v2/coins/tdash.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var dash_1 = require("./dash");
var Tdash = /** @class */ (function (_super) {
    __extends(Tdash, _super);
    function Tdash(bitgo) {
        // TODO: move to @bitgo/utxo-lib (BG-6821)
        return _super.call(this, bitgo, {
            messagePrefix: '\x19Dash Signed Message:\n',
            bip32: {
                public: 0x0488b21e,
                private: 0x0488ade4,
            },
            pubKeyHash: 0x8c,
            scriptHash: 0x13,
            wif: 0xef,
            dustThreshold: 0,
            dustSoftThreshold: 100000,
            feePerKb: 100000,
            coin: 'dash',
        }) || this;
    }
    Tdash.createInstance = function (bitgo) {
        return new Tdash(bitgo);
    };
    Tdash.prototype.getChain = function () {
        return 'tdash';
    };
    Tdash.prototype.getFullName = function () {
        return 'Testnet Dash';
    };
    return Tdash;
}(dash_1.Dash));
exports.Tdash = Tdash;
