// bitgo/dist/src/v2/recovery/blockchairApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("superagent");
var BlockchairCoin = [
    'bitcoin',
    'bitcoin-sv',
    'bitcoin-cash',
];
var devBase = ['dev', 'latest', 'local', 'localNonSecure', 'adminDev', 'adminLatest'];
var testnetBase = ['test', 'adminTest'];
var mainnetBase = ['prod', 'staging', 'adminProd'];
var BlockchairApi = /** @class */ (function () {
    function BlockchairApi(bitgo, coin, apiToken) {
        if (!BlockchairCoin.includes(coin)) {
            throw new Error("coin " + coin + " not supported by blockchair");
        }
        this.bitgo = bitgo;
        this.coin = coin;
        this.apiToken = apiToken;
    }
    BlockchairApi.getBaseUrl = function (env, coin) {
        var url;
        if (mainnetBase.includes(env)) {
            url = "https://api.blockchair.com/" + coin;
        }
        else if (testnetBase.includes(env) || devBase.includes((env))) {
            url = "https://api.blockchair.com/" + coin + "/testnet";
        }
        else if (env === 'mock') {
            url = 'https://api.blockchair.fakeurl/${coin}/testnet';
        }
        else {
            throw new Error("Environment " + env + " unsupported");
        }
        return url;
    };
    /** @inheritdoc */
    BlockchairApi.prototype.getExplorerUrl = function (query) {
        var env = this.bitgo.getEnv();
        var baseUrl = BlockchairApi.getBaseUrl(env, this.coin);
        if (this.apiToken) {
            return baseUrl + query + ("?key=" + this.apiToken);
        }
        return baseUrl + query;
    };
    /** @inheritDoc */
    BlockchairApi.prototype.getAccountInfo = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // we are using blockchair api: https://blockchair.com/api/docs#link_300
                        // https://api.blockchair.com/{:btc_chain}/dashboards/address/{:address}₀
                        if (!address || address.length === 0) {
                            throw new Error('invalid address');
                        }
                        return [4 /*yield*/, request.get(this.getExplorerUrl("/dashboards/address/" + address))];
                    case 1:
                        response = _a.sent();
                        return [2 /*return*/, {
                                txCount: response.body.data[address].address.transaction_count,
                                totalBalance: response.body.data[address].address.balance,
                            }];
                }
            });
        });
    };
    /** @inheritDoc */
    BlockchairApi.prototype.getUnspents = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var response, rawUnspents;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        // using blockchair api: https://blockchair.com/api/docs#link_300
                        // https://api.blockchair.com/{:btc_chain}/dashboards/address/{:address}₀
                        // example utxo from response:
                        // {block_id":-1,"transaction_hash":"cf5bcd42c688cb7c55b5811645e7f0d2a000a85564ca3d6b9fc20f57e14b30bb","index":1,"value":558},
                        if (!address || address.length === 0) {
                            throw new Error('invalid address');
                        }
                        return [4 /*yield*/, request.get(this.getExplorerUrl("/dashboards/address/" + address))];
                    case 1:
                        response = _a.sent();
                        rawUnspents = response.body.data[address].utxo;
                        return [2 /*return*/, rawUnspents.map(function (unspent) {
                                return {
                                    amount: unspent.value,
                                    n: unspent.index,
                                    txid: unspent.transaction_hash,
                                    address: address,
                                };
                            })];
                }
            });
        });
    };
    return BlockchairApi;
}());
exports.BlockchairApi = BlockchairApi;
