// bitgo/dist/src/v2/recovery/blockstreamApi.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var common = require("../../common");
var request = require("superagent");
var errors_1 = require("../../errors");
var BlockstreamApi = /** @class */ (function () {
    function BlockstreamApi(bitgo, apiToken) {
        this.bitgo = bitgo;
        this.apiToken = apiToken;
    }
    /** @inheritDoc */
    BlockstreamApi.prototype.getExplorerUrl = function (query) {
        if (this.apiToken) {
            // TODO: Blockstream does not require an API key for now, howeveer, at some point they may
        }
        return common.Environments[this.bitgo.getEnv()].blockstreamBaseUrl + query;
    };
    /** @inheritDoc */
    BlockstreamApi.prototype.getAccountInfo = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var response, totalBalance, e_1, errorMessage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, request.get(this.getExplorerUrl("/address/" + address))];
                    case 1:
                        response = _a.sent();
                        totalBalance = response.body.chain_stats.funded_txo_sum - response.body.chain_stats.spent_txo_sum;
                        return [2 /*return*/, {
                                txCount: response.body.chain_stats.tx_count,
                                totalBalance: totalBalance,
                            }];
                    case 2:
                        e_1 = _a.sent();
                        errorMessage = "Failed to get address information for " + address + " from " + this.getExplorerUrl('');
                        errorMessage += (e_1.response.status) ? " - " + e_1.response.status : '';
                        errorMessage += (e_1.response.text) ? ": " + e_1.response.text : '';
                        throw new errors_1.BlockExplorerUnavailable(errorMessage);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    /** @inheritDoc */
    BlockstreamApi.prototype.getUnspents = function (address) {
        return __awaiter(this, void 0, void 0, function () {
            var response, rawUnspents, e_2, errorMessage;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 2, , 3]);
                        return [4 /*yield*/, request.get(this.getExplorerUrl("/address/" + address + "/utxo"))];
                    case 1:
                        response = _a.sent();
                        rawUnspents = response.body;
                        return [2 /*return*/, rawUnspents.map(function (unspent) {
                                return {
                                    amount: unspent.value,
                                    n: unspent.vout,
                                    txid: unspent.txid,
                                    address: address,
                                };
                            })];
                    case 2:
                        e_2 = _a.sent();
                        errorMessage = "Failed to get unspents information for " + address + " from " + this.getExplorerUrl('');
                        errorMessage += (e_2.response.status) ? " - " + e_2.response.status : '';
                        errorMessage += (e_2.response.text) ? ": " + e_2.response.text : '';
                        throw new errors_1.BlockExplorerUnavailable(errorMessage);
                    case 3: return [2 /*return*/];
                }
            });
        });
    };
    return BlockstreamApi;
}());
exports.BlockstreamApi = BlockstreamApi;
