// @bitgo/account-lib/dist/src/coin/baseCoin/enum.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Internal metadata. Defines the type of transaction.
 */
var TransactionType;
(function (TransactionType) {
    TransactionType[TransactionType["Send"] = 0] = "Send";
    // Initialize a wallet on-chain (e.g. Multi-sig contract deployment)
    TransactionType[TransactionType["WalletInitialization"] = 1] = "WalletInitialization";
    // Initialize an address on-chain(e.g. Forwarder contract deployment)
    TransactionType[TransactionType["AddressInitialization"] = 2] = "AddressInitialization";
    // Flush tokens from a forwarder address to its base address
    TransactionType[TransactionType["FlushTokens"] = 3] = "FlushTokens";
    // Send a raw single-sig transaction
    TransactionType[TransactionType["SingleSigSend"] = 4] = "SingleSigSend";
    // Update an account on-chain (e.g. Public key revelation operation for Tezos)
    TransactionType[TransactionType["AccountUpdate"] = 5] = "AccountUpdate";
    // Lock
    TransactionType[TransactionType["StakingLock"] = 6] = "StakingLock";
    // Vote
    TransactionType[TransactionType["StakingVote"] = 7] = "StakingVote";
    // Unvote
    TransactionType[TransactionType["StakingUnvote"] = 8] = "StakingUnvote";
    // Activate
    TransactionType[TransactionType["StakingActivate"] = 9] = "StakingActivate";
    // Unlock
    TransactionType[TransactionType["StakingUnlock"] = 10] = "StakingUnlock";
    // Withdraw
    TransactionType[TransactionType["StakingWithdraw"] = 11] = "StakingWithdraw";
})(TransactionType = exports.TransactionType || (exports.TransactionType = {}));
/**
 * Generic list of encoding formats. Can be used as arguments for methods inputs.
 */
var AddressFormat;
(function (AddressFormat) {
    AddressFormat["hex"] = "hex";
    AddressFormat["base58"] = "base58";
})(AddressFormat = exports.AddressFormat || (exports.AddressFormat = {}));
var StakingOperationTypes;
(function (StakingOperationTypes) {
    StakingOperationTypes[StakingOperationTypes["LOCK"] = 0] = "LOCK";
    StakingOperationTypes[StakingOperationTypes["VOTE"] = 1] = "VOTE";
    StakingOperationTypes[StakingOperationTypes["UNVOTE"] = 2] = "UNVOTE";
    StakingOperationTypes[StakingOperationTypes["ACTIVATE"] = 3] = "ACTIVATE";
    StakingOperationTypes[StakingOperationTypes["UNLOCK"] = 4] = "UNLOCK";
    StakingOperationTypes[StakingOperationTypes["WITHDRAW"] = 5] = "WITHDRAW";
})(StakingOperationTypes = exports.StakingOperationTypes || (exports.StakingOperationTypes = {}));
