// @bitgo/account-lib/dist/src/coin/baseCoin/secp256k1ExtendedKeyPair.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var utxo_lib_1 = require("@bitgo/utxo-lib");
var Crypto = __importStar(require("../../utils/crypto"));
var errors_1 = require("./errors");
/**
 * Base class for SECP256K1 extended keypairs.
 */
var Secp256k1ExtendedKeyPair = /** @class */ (function () {
    /**
     * Public constructor. By default, creates a key pair with a random master seed.
     *
     * @param {KeyPairOptions} source Either a master seed, a private key (extended or raw), or a public key
     *     (extended, compressed, or uncompressed)
     */
    function Secp256k1ExtendedKeyPair(source) {
        this.source = source;
    }
    /**
     * Build a Hierarchical Deterministic node or an ECPair from a private key.
     *
     * @param {string} prv An extended or raw private key
     */
    Secp256k1ExtendedKeyPair.prototype.recordKeysFromPrivateKey = function (prv) {
        if (Crypto.isValidXprv(prv)) {
            this.hdNode = utxo_lib_1.HDNode.fromBase58(prv);
        }
        else if (Crypto.isValidPrv(prv)) {
            // Cannot create the HD node without the chain code, so create a regular Key Chain
            this.keyPair = utxo_lib_1.ECPair.fromPrivateKeyBuffer(Buffer.from(prv, 'hex'));
        }
        else {
            throw new Error('Unsupported private key');
        }
    };
    /**
     * Build a Hierarchical Deterministic node or an ECPair from a public key.
     *
     * @param {string} pub - An extended, compressed, or uncompressed public key
     */
    Secp256k1ExtendedKeyPair.prototype.recordKeysFromPublicKey = function (pub) {
        if (Crypto.isValidXpub(pub)) {
            this.hdNode = utxo_lib_1.HDNode.fromBase58(pub);
        }
        else if (Crypto.isValidPub(pub)) {
            // Cannot create an HD node without the chain code, so create a regular Key Chain
            this.keyPair = utxo_lib_1.ECPair.fromPublicKeyBuffer(Buffer.from(pub, 'hex'));
        }
        else {
            throw new Error('Unsupported public key: ' + pub);
        }
    };
    /**
     * Get the extended public key, and the private key if one is available. This is only possible
     * when the key pair was created from a seed or extended keys.
     *
     * @returns {ExtendedKeys} The Extended keys object
     */
    Secp256k1ExtendedKeyPair.prototype.getExtendedKeys = function () {
        if (!this.hdNode) {
            throw new Error('Cannot get extended keys');
        }
        var result = {
            xpub: this.hdNode.neutered().toBase58(),
        };
        // A neutered HD node means it only contains the public key information
        if (!this.hdNode.isNeutered()) {
            result.xprv = this.hdNode.toBase58();
        }
        return result;
    };
    Secp256k1ExtendedKeyPair.prototype.getAddress = function (format) {
        throw new errors_1.NotImplementedError('getAddress not implemented');
    };
    Secp256k1ExtendedKeyPair.prototype.getKeys = function () {
        throw new errors_1.NotImplementedError('getKeys not implemented');
    };
    return Secp256k1ExtendedKeyPair;
}());
exports.Secp256k1ExtendedKeyPair = Secp256k1ExtendedKeyPair;
